src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+import { DidDocument } from '@decentralized-identity/did-common-typescript';
+import * as Base58 from 'bs58';
+import { Cas } from './Cas';
+import Multihash from './Multihash';
+import { WriteOperation, OperationType } from './Operation';
 import Transaction from './Transaction';
-import { WriteOperation } from './Operation';
 
 /**
- * Interface for caching all the processed DIDs and related states.
- * This interface is mainly useful for creating a mock instances for testing purposes.
+ * VersionId identifies the version of a DID document. We use the hash of the
+ * operation that produces a particular version of a DID document as its versionId.
+ * This usage is guaranteed to produce unique VersionId's since the operation contains
+ * as one of its properties the previous VersionId. Since the operation hash is
+ * just a string we alias VersionId to string.
+ *
+ * With this usage, the operation hash serves two roles (1) an identifier for an operation
+ * (2) an identifier for the DID document produced by the operation. In the code below,
+ * we always use VersionId in places where we mean (2) and an OperationHash defined below
+ * when we mean (1).
  */
-export interface DidCache {
-  /** The transaction that was COMPLETELY processed. */
-  readonly lastProcessedTransaction?: Transaction;
+export type VersionId = string;
+
+/**
+ * Alias OperationHash to string - see comment above
+ */
+export type OperationHash = string;
+
+/**
+ * Function type that updates a Did document given an operation. This would be instantiated
+ * with a function that implements json patch application. Using the interface instead of
+ * the actual function hides details of json patching from Did cache.
+ */
+export interface DidDocumentUpdater {
+  (didDoc: DidDocument, operation: WriteOperation): DidDocument;
+}
+
+/**
+ * Function type that creates a Did document given a create/recover operation.
+ */
+export interface DidDocumentCreator {
+  (createOp: WriteOperation): DidDocument;
+}
+
+/**
+ * The timestamp of an operation. We define a linear ordering of
+ * timestamps using the function lesser below.
+ */
+interface OperationTimestamp {
+  readonly transactionNumber: number;
+  readonly operationIndex: number;
+}
+
+function lesser (ts1: OperationTimestamp, ts2: OperationTimestamp): boolean {
+  return ((ts1.transactionNumber < ts2.transactionNumber) ||
+          (ts1.transactionNumber === ts2.transactionNumber) && (ts1.operationIndex < ts2.operationIndex));
+}
+
+/**
+ * Information about a write operation relevant for the DID cache, a subset of the properties exposed by
+ * WriteOperation.
+ */
+interface OperationInfo extends OperationTimestamp {
+  readonly batchFileHash: string;
+  readonly type: OperationType;
+}
+
+/**
+ * DIDCache is a singleton class whose instance holds most of the state in Sidetree node.
```

Comment:
Nit: fix Casing: DidCache. There is no singleton pattern applied here, so consider rephrase the sentence.

Reply:
Fixed.